alea/src/batch.cpp

```diff
@@ -0,0 +1,218 @@
+#include <alps/alea/batch.hpp>
+#include <alps/alea/variance.hpp>
+#include <alps/alea/covariance.hpp>
+#include <alps/alea/internal/util.hpp>
+
+#include <numeric>
+#include <iostream>
+
+namespace alps { namespace alea {
+
+template <typename T>
+batch_data<T>::batch_data(size_t size, size_t num_batches)
+    : batch_(size, num_batches)
+    , count_(num_batches)
+{
+    reset();
+}
+
+template <typename T>
+void batch_data<T>::reset()
+{
+    batch_.fill(0);
+    count_.fill(0);
+}
+
+template class batch_data<double>;
+template class batch_data<std::complex<double> >;
+
+
+template <typename T>
+batch_acc<T>::batch_acc()
+    : size_(-1)
+    , num_batches_(0)
+    , base_size_(0)
+    , store_()
+    , cursor_(0)
+    , offset_()
+{ }
+
+
+template <typename T>
+batch_acc<T>::batch_acc(size_t size, size_t num_batches, size_t base_size)
+    : size_(size)
+    , num_batches_(num_batches)
+    , base_size_(base_size)
+    , store_(new batch_data<T>(size, num_batches))
+    , cursor_(num_batches)
+    , offset_(num_batches)
+{
+    if (num_batches % 2 != 0) {
+        throw std::runtime_error("Number of batches must be even to allow "
+                                 "for rebatching.");
+    }
+    for (size_t i = 0; i != num_batches; ++i)
+        offset_[i] = i * base_size_;
+}
+
+template <typename T>
+batch_acc<T>::batch_acc(const batch_acc &other)
+    : size_(other.size_)
+    , num_batches_(other.num_batches_)
+    , base_size_(other.base_size_)
+    , store_(other.store_ ? new batch_data<T>(*other.store_) : NULL)
+    , cursor_(other.cursor_)
+    , offset_(other.offset_)
+{ }
+
+template <typename T>
+batch_acc<T> &batch_acc<T>::operator=(const batch_acc &other)
+{
+    size_ = other.size_;
+    num_batches_ = other.num_batches_;
+    base_size_ = other.base_size_;
+    store_.reset(other.store_ ? new batch_data<T>(*other.store_) : NULL);
+    cursor_ = other.cursor_;
+    offset_ = other.offset_;
+    return *this;
+}
+
+template <typename T>
+void batch_acc<T>::reset()
+{
+    internal::check_init(*this);
+
+    cursor_.reset();
+    for (size_t i = 0; i != num_batches_; ++i)
+        offset_[i] = i * base_size_;
+
+    if (valid())
+        store_->reset();
+    else
+        store_.reset(new batch_data<T>(size_, num_batches_));
+}
+
+template <typename T>
+batch_acc<T> &batch_acc<T>::operator<<(const computed<T> &source)
+{
+    internal::check_valid(*this);
+
+    // batch is full, move the cursor.
+    // Doing this before the addition ensures no empty batches.
+    if (store_->count()(cursor_.current()) == current_batch_size())
+        next_batch();
+
+    // Since Eigen matrix are column-major, we can just pass the pointer
+    source.add_to(sink<T>(store_->batch().col(cursor_.current()).data(), size()));
+    store_->count()(cursor_.current()) += 1;
+
+    return *this;
+}
+
+template <typename T>
+void batch_acc<T>::next_batch()
+{
+    ++cursor_;
```

Comment:
Why in some methods you check for validness, and here for example you don't

Reply:
`next_batch` is not visible, that's why there is no check here.